parser/README.md

```diff
@@ -0,0 +1,134 @@
+# Integrating with Parsers of SQL Engines
+
+## The Challenge
+
+Typical SQLFlow statements are like the following.
+
+```sql
+SELECT * FROM training_data TRAIN DNNClassifier LABEL kind INTO my_model;
+
+SELECT * FROM testing_data PREDICT testing_data.predicted_kind USING my_model;
+```
+
+The point is, assuming someone already has a SELECT statement for data cleaning and augmentation, (s)he could add a TRAIN or PREDICT clause to enable AI, no matter how complex the statement is or if it is hundreds of lines of code including nested SELECT.
+
+A significant challenge for SQLFlow here is to parse the SQL. Many SQL engines, such as MySQL, Oracle, Hive, SparkSQL, Flink, claim they are compatible with ANSI SQL but most have unique features.
+
+Think about the following facts that (1) each parser might be very complicated, for example, MySQL's parser file sql_yacc.yy has >16,000 lines of code, and (2) some dialects have contradictory uses of the same keyword, it is intractable to merge multiple dialect parsers in the "SQLFlow parser" and keeps it compatible with the development of the dialects.
+
+An alternative is to re-implement SQLFlow for each SQL engine as their extensions or UDF (user-defined function). This approach is equally intractable as there are so many engines to support.
+
+## Overall Design
+
+We design SQLFlow as a wrapper of these SQL engines, which means that SQL statements come directly to SQLFlow and SQLFlow decides if it follows the dialect syntax of the specified SQL engine. If so, SQLFlow proxies the statement to the SQL engine, otherwise, SQLFlow translates it into a Python program, a.k.a., the submitter program, which calls the SQL engine and the AI engine to train or to predict.
+
+With this design, the key challenge becomes to judge if a SQL statement is acceptable by a SQL engine's parser. We propose the following solution: 
+
+```go
+func Parse(sql string) (acceptable bool, err error) {
+    // Where SQL is acceptable by the original engine, no error occurred.
+    err := sql_engine.Parse(sql)
+    if err == nil {
+        return true, ni
+    }
+    
+    // Error message from a parser should contain error position.
+    pos := parseErrorPosition(err) 
+    leftPart = sql[:pos]
+    rightPart = sql[pos:]
+
+    errLeft := sql_engine.Parse(leftPart)
+    // In this case, the SQL is not acceptable due to the syntax error
+    if err != nil {
+       return false, err 
+    }
+
+    // If leftPart is acceptable, it is a legitimate  SELECT statement. We then try right part with SQLFlow parser. 
+    errRight := SQLFlow.Parse(rightPart)
```

Comment:
I would suggest putting the right part before the left part to reduce several network latency. 
Because of the `SQLFlow.Parse` run **locally** always but `sql_engine.Parse` not.